blueocean-indexing/src/main/java/io/jenkins/blueocean/indexing/Index.java

```diff
@@ -0,0 +1,223 @@
+     * @throws IOException
+     */
+    public void addDocuments(Iterable<T> items, Function<T, Document> transformer) throws IOException {
+        try (StandardAnalyzer analyzer = new StandardAnalyzer()) {
+            IndexWriterConfig config = new IndexWriterConfig(analyzer);
+            try (IndexWriter writer = new IndexWriter(index, config)) {
+                writer.addDocuments(Iterables.transform(items, transformer));
+            }
+        }
+    }
+
+    public Iterable<T> query(Query query, int limit, @Nullable Sort sort, final Function<Document, T> transformer) throws IOException {
+        try (IndexReader reader = DirectoryReader.open(index)) {
+            final IndexSearcher searcher = new IndexSearcher(reader);
+            final TopDocs docs = searcher.search(query, limit, sort);
+            final List<ScoreDoc> scoreDocs = Arrays.asList(docs.scoreDocs);
+            return ImmutableList.copyOf(Iterables.transform(scoreDocs, new Function<ScoreDoc, T>() {
+                @Override
+                public T apply(ScoreDoc input) {
+                    int documentId = input.doc;
+                    try {
+                        Document document = searcher.doc(documentId);
+                        return transformer.apply(document);
+                    } catch (IOException e) {
+                        throw new IllegalStateException("Could not load document <" + documentId + "> from index <" + index.toString() + ">", e);
+                    }
+                }
+            }));
+        }
+    }
+
+    public ClosableIterator<Supplier<T>> pagedQuery(final Query query, @Nullable final Sort sort, final Function<Document, T> transformer) throws IOException {
+        final QueryIterator queryIterator = new QueryIterator(query, sort);
+        return new ClosableIterator<Supplier<T>>() {
+            @Override
+            public boolean hasNext() {
+                return queryIterator.hasNext();
+            }
+
+            @Override
+            public Supplier<T> next() {
+                // Wrap this in a Supplier so the caller can choose when to load the document within
+                return new Supplier<T>() {
+                    @Override
+                    public T get() {
+                        ScoreDoc next = queryIterator.next();
+                        Document document = null;
+                        try {
+                            document = queryIterator.searcher.doc(next.doc);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                        return transformer.apply(document);
+                    }
+                };
+            }
+
+            @Override
+            public void close() throws IOException {
+                queryIterator.close();
+            }
+        };
+    }
+
+    public interface ClosableIterator<I> extends Iterator<I>, Closeable {}
```

Comment:
BOs paging code would have to call `close` somewhere up the stack of the iterator supports closing